plugins/widget/plugin.js

```diff
@@ -2901,6 +2944,82 @@
 		} );
 	}
 
+	function handleStylesOnDowncast( widget, widgetElement ) {
+		var styledElement;
+
+		if ( widgetElement ) {
+			if ( widget.styleDefinition && widgetElement.name !== widget.styleDefinition.element ) {
+				styledElement = widgetElement.getFirst( widget.styleDefinition.element );
+			}
+			if ( !styledElement ) {
+				styledElement = widgetElement;
+			}
+
+			styledElement.attributes = styledElement.attributes || {};
+			if ( widget.styleDefinition && widget.styleDefinition.styles ) {
+				styledElement.attributes.style = CKEDITOR.tools.writeCssText( widget.styleDefinition.styles );
+			}
+		}
+	}
+
+	function handleStylesOnUpcast( element, wrapper ) {
+		var style = element.attributes.style,
+			lockedStyle,
+			outputStyle = {},
+			key,
+			styleDefinition = {};
+
+		styleDefinition = styleDefinition || { attributes: {}, lockedStyle: {} };
+		styleDefinition.attributes = styleDefinition.attributes || {};
+		styleDefinition.element = element.name;
+		if ( element.lockedStyle ) {
+			lockedStyle = element.lockedStyle;
+			delete element.lockedStyle;
+		} else {
+			lockedStyle = {};
+		}
+
+		for ( key in element.attributes ) {
+			if ( key !== 'style' && key.substring( 0, 4 ).toLowerCase() !== 'data' ) {
+				styleDefinition.attributes[ key ] = element.attributes[ key ];
+			}
+		}
+
+		// If element is <a> we want to keep styles so it can be changed via link dialog.
+		if ( styleDefinition.element === 'a' ) {
+			style = CKEDITOR.tools.parseCssText( style );
+			for ( key in style ) {
+				lockedStyle[ key ] = style[ key ];
+			}
+			style = CKEDITOR.tools.writeCssText( style );
+		}
+
+		// Preserve styles that are set as 'locked', it is important so other plugins extending widget can prevent moving some styles to wrapper.
+		if ( CKEDITOR.tools.objectKeys( lockedStyle ).length && CKEDITOR.tools.objectKeys( style ).length ) {
+			style = CKEDITOR.tools.parseCssText( style );
+			for ( key in style ) {
+				if ( !( key in lockedStyle ) ) {
+					outputStyle[ key ] = style[ key ];
+					delete style[ key ];
+				}
+			}
+			style = CKEDITOR.tools.writeCssText( style );
+			element.attributes.style = style;
+			style = CKEDITOR.tools.writeCssText( outputStyle );
+		} else {
+			// When we don't preserve any styles we can delete them on widget.element
+			delete element.attributes.style;
+		}
+		if ( style ) {
+			wrapper.attributes.style = style;
+			styleDefinition.styles = CKEDITOR.tools.parseCssText( style );
+		} else {
+			// If there are no styles remove it to make sure we don't have `styles=''` in our output.
+			delete styleDefinition.styles;
+		}
+		wrapper.attributes[ 'data-style-definition' ] = JSON.stringify( styleDefinition );
```

Comment:
This might be considered as 'ugly' solution, but I couldn't find better way to store `widget.styleDefinition` before widget is initialized.

Reply:
At least it should be internal CKE attribute, `[data**-cke**-style-definition]`.

What's more, information about partial styles (classes) are inside widget's data. Maybe we should reuse this mechanism and include all styles there.